cpp/src/experimental/louvain.cuh

```diff
@@ -33,21 +365,1567 @@ class Louvain {
+    //
+    thrust::sequence(rmm::exec_policy(stream_)->on(stream_),
+                     cluster_v_.begin(),
+                     cluster_v_.end(),
+                     base_vertex_id_);
+    thrust::copy(
+      rmm::exec_policy(stream_)->on(stream_), cluster_v_.begin(), cluster_v_.end(), d_cluster_vec);
+
+    while (num_level < max_level) {
+      compute_vertex_and_cluster_weights();
+
+      weight_t new_Q = update_clustering(total_edge_weight, resolution);
+
+#ifdef DEBUG
+      std::cout << "new_Q = " << new_Q << std::endl;
+#endif
+
+      if (new_Q <= best_modularity) { break; }
+
+      best_modularity = new_Q;
+
+      shrink_graph(d_cluster_vec);
+
+      num_level++;
+    }
+
+    timer_display(std::cout);
+
+#ifdef DEBUG
+    print_v("resulting cluster", d_cluster_vec, number_of_vertices_);
+#endif
+
+    return std::make_pair(num_level, best_modularity);
+  }
+
+ protected:
+  void timer_start(std::string const &region)
+  {
+#ifdef TIMING
+    if (rank_ == 0) hr_timer_.start(region);
+#endif
+  }
+
+  void timer_stop(cudaStream_t stream)
+  {
+#ifdef TIMING
+    if (rank_ == 0) {
+      CUDA_TRY(cudaStreamSynchronize(stream));
+      hr_timer_.stop();
+    }
+#endif
+  }
+
+  void timer_display(std::ostream &os)
+  {
+#ifdef TIMING
+    if (rank_ == 0) hr_timer_.display(os);
+#endif
+  }
+
+ public:
+  // TODO CHECK:  Should be right, as long as
+  //
+  //    src_cluster_cache_v_
+  //    dst_cluster_cache_v_
+  //
+  // are populated properly
+  //
+  weight_t modularity(weight_t total_edge_weight, weight_t resolution)
+  {
+    weight_t Q = -resolution / (total_edge_weight * total_edge_weight) *
+                 experimental::transform_reduce_v(
+                   handle_,
+                   current_graph_view_,
+                   cluster_weights_v_.begin(),
+                   [] __device__(auto p) { return p * p; },
+                   weight_t{0});
+
+#if 1
+    std::cout << "** Q (subtraction) = " << Q << std::endl;
+
+    print_v("weights", vertex_weights_v_);
+    print_v("src_cluster_cache", src_cluster_cache_v_);
+    print_v("dst_cluster_cache", dst_cluster_cache_v_);
+#endif
+
+    rmm::device_vector<weight_t> increase_v(local_num_vertices_, weight_t{0.0});
+
+    experimental::copy_v_transform_reduce_out_nbr(
+      handle_,
+      current_graph_view_,
+      src_cluster_cache_v_.begin(),
+      dst_cluster_cache_v_.begin(),
+      [] __device__(auto src, auto, auto weight, auto src_cluster, auto nbr_cluster) {
+        if (src_cluster == nbr_cluster) {
+          return weight;
+        } else
+          return weight_t{0};
+      },
+      weight_t{0},
+      increase_v.begin());
+
+#if 1
+    CHECK_CUDA(stream_);
+    print_v("increase", increase_v);
+#endif
+
+    Q += (experimental::transform_reduce_v(
+            handle_,
+            current_graph_view_,
+            increase_v.begin(),
+            [] __device__(auto p) { return p; },
+            weight_t{0}) /
+          total_edge_weight);
```

Comment:
If I am not mistaken, increase_v is just used for summation, can't you merge copy_v_transform_reduce_out_nbr and transform_reduce_v with transform_reduce_e?

Reply:
```
Q += experimental::transform_reduce_e(
  handle_, current_graph_view_, src_cluster, nbr_cluster,
  [] __device__  (auto src, auto, auto weight, auto src_cluster, auto nbr_cluster) {
        if (src_cluster == nbr_cluster) {
          return weight;
        } else
          return weight_t{0};
      },
      weight_t{0});
```
Then, no need for the temporary increase_v